### ISSUE TITLE
Default Project Options Not Updating

### ISSUE THREAD

**Issue by SirNate0** [mesonbuild/meson#13556]
**Describe the bug**
I set `cpp_std=c++17` as a default option for my project. I later decided to change to `c++20`. I cannot seem to get it to actually update the value. The [FAQ](https://mesonbuild.com/FAQ.html#why-are-changes-to-default-project-options-ignored) suggests that I should run `meson configure` to reconfigure your build dir. However, running this seems not to reconfigure the build dir, it simply shows that it still has a value of `c++17`.

**To Reproduce**
```meson
project('foobar', 'cpp', default_options : ['cpp_std=c++14'])
```
1. Create build dir with that (example from FAQ).
2. Update the default value to c++20
3. Run `meson configure` in the build dir

**Expected behavior**
*If we are going with the poor decision opted for in the FAQ*
`meson configure` reconfigures the build dir with the new default, as the FAQ seems to suggest.

*If we want the actual expected behavior*
Meson should save whether values are the default value or not separately from the actual value. If the default changes and the value was marked as using the default, then the value is updated. If the value is not marked as being default, then it is not changed (even if it was the same value as the default). As the FAQ says, "default values are only defaults and will never affect the value of an option **once set**" -- running meson does not constitute setting the value, unless I explicitly set the value.

**system parameters**
* Native build
* Kubuntu 22.04
* Python 3.10.12
* Meson 0.61.2

**Comment by dcbaker** [comment]
The FAQ is very clear there, default_options are the default values read in when Meson is first configuring a build directory. They are the values of those options. This is different than an unset option, which would happen if a new option is added to your `meson.options` file and then you reconfigure. Running meson **does** constitute setting the value.

This is an extremely non-trivial thing to fix, for a lot of reasons. Maybe it will happen with all of the option handling rework going on, but I kinda doubt it.

The thing that you need to do in this case is either configure a new build directory, or use `meson configure` to change it manually.
1. start over: rm -rf builddir; meson setup builddir
2. `meson configure builddir -Dcpp_std=c++20`

**Comment by SirNate0** [comment]
> Running meson does constitute setting the value.

No. Maybe from meson's perspective it does, but it's wrong. I, the user, have not set the value. Pretending that I have is a bug. (If, for some unknown reason, other users make use of this bug, it should be simple enough to add a command that would bake/freeze all of the default options to explicitly set ones).

I hope it does end up being fixed, otherwise it's just one more thing to add to an increasing list of things that make meson unfriendly to work with. Regarding that, it sounds like meson actually already supports reconfiguring default options automatically through these unset (one could say 'marked as default') options: "an unset option, which would happen if a new option is added to your meson.options file and then you reconfigure." Is there a reason this is not how default options are normally handled already?

Regarding workarounds:

1. Removing and recreating the build directory is simply not an acceptable solution. 
2. This should be specified in the docs in a manner like you have here. The FAQ makes it sound like I should just run `meson configure` and it will do the changes. Perhaps it is obvious if you have used meson for a while, but it was not at all apparent to me as a new casual user (especially while tired at midnight).
